src/gluonts/model/deepstate/_estimator.py

```diff
@@ -0,0 +1,312 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+# Standard library imports
+from typing import List, Optional
+
+# Third-party imports
+from mxnet.gluon import HybridBlock
+
+# First-party imports
+from gluonts.core.component import validated
+from gluonts.model.deepstate.issm import ISSM, CompositeISSM
+from gluonts.model.estimator import GluonEstimator
+from gluonts.model.predictor import Predictor, RepresentableBlockPredictor
+from gluonts.support.util import copy_parameters
+from gluonts.time_feature.lag import (
+    longest_period_from_frequency_str,
+    TimeFeature,
+    time_features_from_frequency_str,
+)
+from gluonts.trainer import Trainer
+from gluonts.transform import (
+    AddObservedValuesIndicator,
+    AddAgeFeature,
+    AddTimeFeatures,
+    AsNumpyArray,
+    Chain,
+    CanonicalInstanceSplitter,
+    ExpandDimArray,
+    FieldName,
+    RemoveFields,
+    SetField,
+    TestSplitSampler,
+    Transformation,
+    VstackFeatures,
+)
+
+# Relative imports
+from ._network import DeepStatePredictionNetwork, DeepStateTrainingNetwork
+
+SEASON_INDICATORS_FIELD = "seasonal_indicators"
+
+
+class DeepStateEstimator(GluonEstimator):
+    """
+    Construct a DeepState estimator.
+    
+    This implements the deep state space model described in
+    [Rangapuram et. al. 2018]
+
+    .. [Rangapuram et. al. 2018] Syama S. Rangapuram, Matthias W. Seeger,
+    Jan Gasthaus, Lorenzo Stella, Yuyang Wang, Tim Januschowski.
+    Deep State Space Models for Time Series Forecasting.
+    In NeurIPS 31, pages 7785–7794, 2018.
+    https://papers.nips.cc/paper/8004-deep-state-space-models-for-time-series-forecasting
+
+    Parameters
+    ----------
+    freq
+        Frequency of the data to train on and predict
+    prediction_length
+        Length of the prediction horizon
+    add_trend
+        Flag to indicate whether to include trend component in the
+        state space model
+    past_length
+        This is the length of the training time series;
+        i.e., number of steps to unroll the RNN for before computing predictions.
+        Set this to (at most) the length of the shortest time series in the dataset.
+        (default: None, in which case the training length is set such that at least
+        `num_seasons_to_train` seasons are included in the training.
+        See `num_seasons_to_train`)
+    num_seasons_to_train
```

Comment:
Maybe a better name for this is `num_periods_to_train`? In my idea of naming things, a period is a full cycle of seasons